module/VuFind/src/VuFind/Db/Service/TagService.php

```diff
@@ -778,4 +779,401 @@ public function getStatistics(bool $extended = false): array
+        $dql = 'SELECT MIN(t.id) AS id, ' . $tag . ' AS tag, COUNT(DISTINCT(rt.resource)) AS cnt '
+            . 'FROM ' . $this->getEntityClass(ResourceTags::class) . ' rt '
+            . 'JOIN rt.tag t '
+            . 'JOIN rt.resource r '
+            . 'JOIN ' . $this->getEntityClass(UserResource::class) . ' ur '
+            . 'WITH r.id = ur.resource '
+            . 'WHERE ur.user = :userId AND rt.user = :userId AND ur.list = rt.list ';
+        $parameters = compact('userId');
+        if (null !== $source) {
+            $dql .= 'AND r.source = :source ';
+            $parameters['source'] = $source;
+        }
+        if (null !== $resourceId) {
+            $dql .= 'AND r.recordId = :resourceId ';
+            $parameters['resourceId'] = $resourceId;
+        }
+        if (null !== $listId) {
+            $dql .= 'AND rt.list = :listId ';
+            $parameters['listId'] = $listId;
+        }
+        $dql .= 'GROUP BY t.tag ORDER BY LOWER(t.tag) ';
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $results = $query->getResult();
+        return $results;
+    }
+
+    /**
+     * Get tags assigned to a user list.
+     *
+     * @param int    $listId List ID
+     * @param string $userId User ID to look up (null for no filter).
+     *
+     * @return array
+     */
+    public function getForList($listId, $userId = null)
+    {
+        $tag = $this->caseSensitive ? 't.tag' : 'lower(t.tag)';
+
+        $dql = 'SELECT MIN(t.id) AS id, ' . $tag . ' AS tag '
+            . 'FROM ' . $this->getEntityClass(ResourceTags::class) . ' rt '
+            . 'JOIN rt.tag t '
+            . 'WHERE rt.list = :listId AND rt.resource IS NULL ';
+        $parameters  = compact('listId');
+        if ($userId) {
+            $dql .= 'AND rt.user = :userId ';
+            $parameters['userId'] = $userId;
+        }
+
+        $dql .= 'GROUP BY t.tag ORDER BY LOWER(t.tag) ';
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $results = $query->getResult();
+        return $results;
+    }
+
+    /**
+     * Get a list of duplicate tags (this should never happen, but past bugs
+     * and the introduction of case-insensitive tags have introduced problems).
+     *
+     * @return array
+     */
+    public function getDuplicates()
+    {
+        $dql = 'SELECT MIN(t.tag) AS tag, COUNT(t.tag) AS cnt, MIN(t.id) AS id '
+            . 'FROM ' . $this->getEntityClass(Tags::class) . ' t '
+            . 'GROUP BY ' . $this->caseSensitive ? 't.tag ' : 'lower(t.tag) '
+            . 'HAVING COUNT(t.tag) > 1';
+        $query = $this->entityManager->createQuery($dql);
+        $results = $query->getResult();
+        return $results;
+    }
+
+    /**
+     * Support method for fixDuplicateTag() -- merge $source into $target.
+     *
+     * @param string $target Target ID
```

Comment:
These are not strings, are they? I think they're Tag objects. Maybe we should add better typing to all of this code while we're refactoring anyway.